### ISSUE TITLE
Requesting client certificates without configuring a client CA certificate chain

### ISSUE THREAD

**Issue by aaronmdjones** [ARMmbed/mbedtls#506]
The documentation for `mbedtls_ssl_conf_authmode` does not state that a CA chain must be set with `mbedtls_ssl_conf_ca_chain` on the server. If this is not done, mbedTLS will not request a certificate if you set the authmode to `OPTIONAL`.

However, experimentation has determined that you can pass an initialised-but-empty `mbedtls_x509_crt` structure to `mbedtls_ssl_conf_ca_chain` on the server, and it will request and process client certificates as expected.

I am opening this issue to determine whether this behaviour can be relied upon -- our software wants to unconditionally request client certificates, but it does not verify them, or want the library to verify them -- it only requests them to generate a fingerprint for them. Naturally, having a hardcoded CA certificate in the application is less than ideal, and the server certificate cannot always be used for this purpose instead either.

Ofcourse, the best approach would be to request client certificates if the authmode is set to `OPTIONAL`, regardless of whether a CA chain has been configured or not. Failing that, can we atleast get it documented whether or not an empty `mbedtls_x509_crt` structure (i.e. one passed to `mbedtls_x509_crt_init` and nothing else) is acceptable for `mbedtls_ssl_conf_ca_chain` on the server side? At the moment, that is the case, but you may consider that a bug.

Thanks.

**Comment by aaronmdjones** [comment]
Further exploration has determined a possible solution:

The offending code is located at https://github.com/ARMmbed/mbedtls/blob/master/library/ssl_tls.c#L4389-L4411

The solution would be to also check on L4407 whether this is a client endpoint, before returning the error. Then, a custom application-provided verify callback could make L4416 return a success condition on the server, and negotiation would succeed.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-822

**Comment by billybednar** [comment]
It appears that using an empty `mbedtls_x509_crt` structure for the ca_chain is safe in 2.2.1, but it certainly doesn't seem like something that would be expected to work in general. As far as I can tell an empty certificate isn't a valid certificate and passing one to anything expecting a certificate is an error, even if it happens works with this particular version. 

In 2.2.1 the `ca_chain` configuration field is read in two places:

1. In `ssl_write_certificate_request` the [chain is used][1] to build the `certificate_authorities` field for the certificate request record. The `version` field of the empty certificate is 0 so the loop is skipped and the certificate isn't otherwise used. If check wasn't there then the `memcpy` would be undefined behavior and the `certificate_authorities` field wouldn't be correctly formed.
2. In `mbedtls_ssl_parse_certificate` the [chain is passed][2] to `mbedtls_x509_crt_verify_with_profile` as `trust_ca` unless you're using `MBEDTLS_SSL_VERIFY_NONE`. It uses it in three places:
    - [Here][5] it's passed to `x509_crt_check_parent`, [which in turn calls `x509_name_cmp`][6]. That [returns -1][7] without calling `memcmp` as `a->oid.tag` is non-zero (it's the client certificate's issuer--the DER parser would have complained if the OID wasn't tagged as an OID) and `b` is all zeroed. That goes back up to the loop, which does not break, and sets `parent = parent->next`, which is NULL.
    - [Calls][8] to [x509_crt_verify_child][9] and [x509_crt_verify_top][10]. If you trace those out you can see that they don't do anything objectionable with the chain. They involve similar calls to `x509_crt_check_parent`.

[1]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/ssl_srv.c#L2580-2604
[2]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/ssl_tls.c#L4389-4422
[3]: https://en.wikipedia.org/wiki/Server_Name_Indication
[4]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/ssl_tls.c#L4415-4421
[5]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/x509_crt.c#L2238-2243
[6]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/x509_crt.c#L1843-1845
[7]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/x509_crt.c#L1806-1812
[8]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/x509_crt.c#L2259-2274
[9]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/x509_crt.c#L2006
[10]: https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.2.1/library/x509_crt.c#L1876

**Comment by mikapak** [comment]
Hi Billybednar,

I am mbedtls 2.3.0 with hard coded CA certificate in my application,  facing the same issue that you have described in point 2. control is not at all coming out of 'x509_crt_verify_child' function. please guide me how to fix this issue.

**Comment by billybednar** [comment]
@mikapak It looks like you're making a client. This issue is about a server side problem where the server doesn't request a certificate from the client unless you provide a chain to validate it against. My analysis above was restricted to seeing if you can safely do
```
mbedtls_x509_crt dummyCaChain;
mbedtls_x509_crt_init(&dummyCaChain);
mbedtls_ssl_conf_ca_chain(&sslConfig, &dummyCaChain, NULL);
```
from a server context as a workaround. I never had any issues with anything getting stuck.

**Comment by mikapak** [comment]
I will check with dummyCA.
Is the CA certificate user readable after it is parsed?. I mean, how to check whether the CA is parsed properly.

**Comment by billybednar** [comment]
@mikapak If a CA certificate is involved then your problem is off-topic.

**Comment by mikapak** [comment]
I  provided the CA by hard coding in my application and parsed using `mbedtls_x509_crt_parse`

**Comment by mikapak** [comment]
Hi Billy Bednar,

I tried with dummyCA, still facing same issue.

**Comment by aaronmdjones** [comment]
Bump: This issue seems to have been inadvertently fixed by commit 39ae8cd2077d

**Comment by RonEld** [comment]
Closing this issue, as it was fixed in #903 , and back-ported in #934 and #935
